File: internal/session/session.go
package session

import (
	"time"

	"github.com/Renderix/freeman/internal/buffer"
	"github.com/Renderix/freeman/internal/engine"
)

// Error codes per spec §3.3
const (
	ErrorCodeInvalidVoice   = "INVALID_VOICE"
	ErrorCodeTtsError       = "TTS_ERROR"
	ErrorCodeBufferOverflow = "BUFFER_OVERFLOW"
	ErrorCodeTimeout        = "TIMEOUT"
	ErrorCodeFatal          = "FATAL"
)

// Session manages an individual TTS streaming session.
type Session struct {
	ID                 string
	Voice              string
	Speed              float64
	Engine             *engine.TTSEngine
	Buffer             *buffer.SentenceBuffer
	SendAudio          func([]byte) error
	SendJSON           func(interface{}) error
	IsActive           bool
	SentenceCount      int
	StartTime          time.Time
	TotalDurationMs    int
	SentencesCompleted int
	PendingSentences   []string
}

// NewSession creates a new session.
func NewSession(id, voice string, speed float64, ttsEngine *engine.TTSEngine, sendAudio func([]byte) error, sendJSON func(interface{}) error) *Session {
	return &Session{
		ID:               id,
		Voice:            voice,
		Speed:            speed,
		Engine:           ttsEngine,
		Buffer:           buffer.NewSentenceBuffer(2.0, 150),
		SendAudio:        sendAudio,
		SendJSON:         sendJSON,
		IsActive:         true,
		StartTime:        time.Now(),
		PendingSentences: make([]string, 0),
	}
}

// HandleText handles incoming text chunks.
func (s *Session) HandleText(chunk string, isFinal bool) (int, error) {
	sentences := s.Buffer.AddChunk(chunk)

	if isFinal {
		finalSentence := s.Buffer.Flush()
		if finalSentence != "" {
			sentences = append(sentences, finalSentence)
		}
	}

	for _, sentence := range sentences {
		if err := s.processSentence(sentence); err != nil {
			return 0, err
		}
	}

	return len(s.PendingSentences), nil
}

func (s *Session) processSentence(text string) error {
	s.SentenceCount++
	sID := s.SentenceCount

	s.PendingSentences = append(s.PendingSentences, text)
	queuePosition := len(s.PendingSentences) - 1

	estimatedDuration := s.estimateDurationMs(text)

	s.SendJSON(map[string]interface{}{
		"type":                   "sentence_start",
		"id":                     sID,
		"text":                   text,
		"estimated_duration_sec": float64(estimatedDuration) / 1000.0,
		"queue_position":         queuePosition,
	})

	startGen := time.Now()

	// Generate audio
	audioBytes, err := s.Engine.Generate(text, s.Voice, s.Speed)

	// Remove from pending
	for i, p := range s.PendingSentences {
		if p == text {
			s.PendingSentences = append(s.PendingSentences[:i], s.PendingSentences[i+1:]...)
			break
		}
	}

	if err == nil && audioBytes != nil {
		if err := s.SendAudio(audioBytes); err != nil {
			return err
		}

		genTimeMs := int(time.Since(startGen).Milliseconds())
		durationMs := s.calculateDurationMs(audioBytes)
		s.TotalDurationMs += durationMs
		s.SentencesCompleted++

		s.SendJSON(map[string]interface{}{
			"type":            "sentence_complete",
			"id":              sID,
			"duration_ms":     durationMs,
			"processing_ms":   genTimeMs,
			"queue_remaining": len(s.PendingSentences),
		})
	} else {
		s.SendJSON(map[string]interface{}{
			"type":        "error",
			"sentence_id": sID,
			"message":     "TTS generation failed",
			"recoverable": true,
			"code":        ErrorCodeTtsError,
		})
	}

	return nil
}

func (s *Session) estimateDurationMs(text string) int {
	// Rough estimate: ~80ms per character
	baseDuration := len(text) * 80
	return int(float64(baseDuration) / s.Speed)
}

func (s *Session) calculateDurationMs(audioBytes []byte) int {
	// WAV header is 44 bytes
	if len(audioBytes) <= 44 {
		return 0
	}
	audioDataBytes := len(audioBytes) - 44
	// 24kHz, 16-bit mono = 48000 bytes per second = 48 bytes per ms
	return int(audioDataBytes / 48)
}

// Flush force flushes the buffer.
func (s *Session) Flush() error {
	sentence := s.Buffer.Flush()
	if sentence != "" {
		return s.processSentence(sentence)
	}
	return nil
}

// CheckTimeout checks for partial sentence timeout.
func (s *Session) CheckTimeout() error {
	sentence := s.Buffer.CheckTimeout()
	if sentence != "" {
		return s.processSentence(sentence)
	}
	return nil
}

// SendStreamEnd sends stream_end message.
func (s *Session) SendStreamEnd(reason string) error {
	return s.SendJSON(map[string]interface{}{
		"type":              "stream_end",
		"total_sentences":   s.SentencesCompleted,
		"total_duration_ms": s.TotalDurationMs,
		"reason":            reason,
	})
}

File: src/session.py
import asyncio
import time
from typing import Optional, Callable, Any, List
from .sentence_buffer import SentenceBuffer


# Error codes per spec §3.3
class ErrorCode:
    INVALID_VOICE = "INVALID_VOICE"
    TTS_ERROR = "TTS_ERROR"
    BUFFER_OVERFLOW = "BUFFER_OVERFLOW"
    TIMEOUT = "TIMEOUT"
    FATAL = "FATAL"


class Session:
    """
    Manages an individual TTS streaming session.
    Orchestrates text buffering, TTS generation, and audio delivery.
    """

    # Audio specs: 24kHz, 16-bit mono = ~48 bytes per ms
    BYTES_PER_MS = 48

    def __init__(
        self,
        session_id: str,
        voice: str,
        speed: float,
        tts_engine: Any,
        send_audio: Callable[[bytes], Any],
        send_json: Callable[[dict], Any]
    ):
        self.id = session_id
        self.voice = voice
        self.speed = speed
        self.tts_engine = tts_engine
        self.send_audio = send_audio
        self.send_json = send_json

        self.buffer = SentenceBuffer()
        self.sentence_count = 0
        self.is_active = True

        # Performance tracking
        self.start_time = time.time()
        self.total_duration_ms = 0

        # Queue tracking for spec-compliant messages
        self.pending_sentences: List[str] = []
        self.sentences_completed = 0

    async def handle_text(self, chunk: str, is_final: bool = False) -> int:
        """Handle incoming text chunk. Returns number of pending sentences."""
        sentences = self.buffer.add_chunk(chunk)

        if is_final:
            final_sentence = self.buffer.flush()
            if final_sentence:
                sentences.append(final_sentence)

        for sentence in sentences:
            await self._process_sentence(sentence)

        return len(self.pending_sentences)

    def _estimate_duration_ms(self, text: str) -> int:
        """Estimate audio duration in ms based on text length and speed."""
        # Rough estimate: ~80ms per character at normal speed
        base_duration = len(text) * 80
        return int(base_duration / self.speed)

    def _calculate_duration_ms(self, audio_bytes: bytes) -> int:
        """Calculate actual audio duration from WAV bytes."""
        # WAV header is 44 bytes, rest is audio data
        # 24kHz, 16-bit mono = 48000 bytes per second
        if len(audio_bytes) <= 44:
            return 0
        audio_data_bytes = len(audio_bytes) - 44
        return int(audio_data_bytes / self.BYTES_PER_MS)

    async def _process_sentence(self, text: str):
        """Generate and send audio for a sentence."""
        self.sentence_count += 1
        s_id = self.sentence_count

        # Add to pending queue
        self.pending_sentences.append(text)
        queue_position = len(self.pending_sentences) - 1

        estimated_duration = self._estimate_duration_ms(text)

        await self.send_json({
            "type": "sentence_start",
            "id": s_id,
            "text": text,
            "estimated_duration_sec": estimated_duration / 1000.0,
            "queue_position": queue_position
        })

        start_gen = time.time()

        # Run TTS in executor
        loop = asyncio.get_event_loop()
        audio_bytes = await loop.run_in_executor(
            None,
            lambda: self.tts_engine.generate(text, self.voice, self.speed)
        )

        # Remove from pending queue
        if text in self.pending_sentences:
            self.pending_sentences.remove(text)

        if audio_bytes:
            await self.send_audio(audio_bytes)

            gen_time_ms = int((time.time() - start_gen) * 1000)
            duration_ms = self._calculate_duration_ms(audio_bytes)
            self.total_duration_ms += duration_ms
            self.sentences_completed += 1

            await self.send_json({
                "type": "sentence_complete",
                "id": s_id,
                "duration_ms": duration_ms,
                "processing_ms": gen_time_ms,
                "queue_remaining": len(self.pending_sentences)
            })
        else:
            await self.send_json({
                "type": "error",
                "sentence_id": s_id,
                "message": "TTS generation failed",
                "recoverable": True,
                "code": ErrorCode.TTS_ERROR
            })

    async def flush(self):
        """Force flush remaining buffer."""
        sentence = self.buffer.flush()
        if sentence:
            await self._process_sentence(sentence)

    async def check_timeout(self):
        """Check for partial sentence timeout."""
        sentence = self.buffer.check_timeout()
        if sentence:
            await self._process_sentence(sentence)

    async def send_stream_end(self, reason: str = "client_end"):
        """Send stream_end message per spec §3.3."""
        await self.send_json({
            "type": "stream_end",
            "total_sentences": self.sentences_completed,
            "total_duration_ms": self.total_duration_ms,
            "reason": reason
        })
